apps/src/lib/script-editor/StageCard.jsx

```diff
@@ -0,0 +1,187 @@
+    });
+    this.setState({
+      drag: position,
+      dragHeight: this.metrics(position).height + levelTokenMargin,
+      initialPageY: pageY,
+      initialScroll: document.body.scrollTop,
+      newPosition: position,
+      startingPositions
+    });
+    window.addEventListener('selectstart', this.preventSelect);
+    window.addEventListener('mousemove', this.handleDrag);
+    window.addEventListener('mouseup', this.handleDragStop);
+  },
+
+  handleDrag({pageY}) {
+    const scrollDelta = document.body.scrollTop - this.state.initialScroll;
+    const delta = pageY - this.state.initialPageY;
+    const dragPosition = this.metrics(this.state.drag).top + scrollDelta;
+    let newPosition = this.state.drag;
+    const currentPositions = this.state.startingPositions.map((midpoint, index) => {
+      const position = index + 1;
+      if (position === this.state.drag) {
+        return delta;
+      }
+      if (position < this.state.drag && dragPosition < midpoint) {
+        newPosition--;
+        return this.state.dragHeight;
+      }
+      if (position > this.state.drag && dragPosition + this.state.dragHeight > midpoint) {
+        newPosition++;
+        return -this.state.dragHeight;
+      }
+      return 0;
+    });
+    this.setState({currentPositions, newPosition});
+  },
+
+  handleDragStop() {
+    if (this.state.drag !== this.state.newPosition) {
+      this.props.reorderLevel(
+        this.props.stage.position,
+        this.state.drag,
+        this.state.newPosition
+      );
+    }
+    this.setState({drag: null, newPosition: null, currentPositions: []});
+    window.removeEventListener('selectstart', this.preventSelect);
+    window.removeEventListener('mousemove', this.handleDrag);
+    window.removeEventListener('mouseup', this.handleDragStop);
+  },
+
+  handleAddLevel() {
+    this.props.addLevel(this.props.stage.position);
+  },
+
+  handleLockableChanged() {
+    this.props.setStageLockable(this.props.stage.position, this.refs.lockable.checked);
+  },
+
+  preventSelect(e) {
+    e.preventDefault();
+  },
+
+  render() {
+    return (
+      <div style={styles.stageCard}>
+        <div style={styles.stageCardHeader}>
+          Stage {this.props.stage.position}: {this.props.stage.name}
+          <OrderControls
+            type={ControlTypes.Stage}
+            position={this.props.stage.position}
+            total={this.props.stagesCount}
+          />
+          <div style={styles.stageLockable}>
+            Require teachers to unlock this stage before students in their
+            section can access it
+            <input
+              ref="lockable"
+              defaultChecked={this.props.stage.lockable}
+              onChange={this.handleLockableChanged}
+              type="checkbox"
+              style={styles.checkbox}
+            />
+          </div>
+        </div>
+        {this.props.stage.levels.map(level =>
+          <LevelToken
+            ref={`levelToken${level.position}`}
```

Comment:
In the first iteration, I think Paul made the comment that string refs are considered legacy (which I was somewhat dismissive of). I wonder tho if this is a case where we could actually benefit from non-string refs. I think the way to do this would be to do something like:
```js
ref={levelToken => this.metrics[level.position] = levelToken.getClientBoundingRect()}
// then elsewhere you can do
this.metrics[token]; // instead of this.metrics(token)
```

(I don't feel that strongly about this, but it might be good for us to start getting away from using string refs).